src/backend/parser/parse_relation.c

```diff
@@ -592,19 +592,35 @@ scanRTEForColumn(ParseState *pstate, RangeTblEntry *rte, char *colname,
 	 */
 	if (rte->rtekind == RTE_RELATION)
 	{
-		/* In GPDB, system columns like gp_segment_id, ctid, xmin/xmax seem to be
-		 * ambiguous for replicated table, replica in each segment has different
-		 * value of those columns, between sessions, different replicas are choosen
-		 * to provide data, so it's weird for users to see different system columns
-		 * between sessions. So for replicated table, we don't expose system columns
-		 * unless it's GP_ROLE_UTILITY for debug purpose.
+		attnum = specialAttNum(colname);
+
+		if (relstorage_is_ao(get_rel_relstorage(rte->relid)) &&
```

Comment:
I think, check for attnum is cheaper than call get_rel_relstorage

Reply:
perhaps, but how is it related to commit logic?

Reply:
put check for attnum first and get_rel_relstorage last